rfcs/framework-0040-relations.md

```diff
@@ -0,0 +1,321 @@
+- Lack of fine grained control. It is not easily possible to control which fields from the referenced record is included. For instance, a biblographic record relates to a grant record that relates to a funder record. All three records will be dereferenced, causing the index JSON to become very big.
+- Security/performance - because JSONRef is based on making HTTP requests, there's a risk that badly formed URLs, will cause HTTP requests to be made. At best this will only cause performance issues, at worst it opens a backdoor for injecting content in the JSON document. While this could also be considered a feature, ability to relate to externally hosted records, we think the performance/security issues makes this technically impossible.
+
+
+## Design
+
+### Challenges
+
+The primary challenges of record relationships includes:
+
+- **Syntax**: How should a relation be made between two records?
+- **Integrity**: How is integrity checked and validated?
+- **Dereferencing**: When and how is a record dereferenced?
+- **Denormalization**: When a dereferenced record is updated, how is the related index updated?
+
+Each of above challenges is addressed in the following design.
+
+#### Syntax
+
+We propose the following syntax:
+
+```
+{
+    "<field>": {"id": "<persistent id>"},
+    "<field>": [{"id": "<persistent id>"}, ...]
+}
+```
+
+A relation is an object with a single key ``id`` with a string value for a persistent identifier for the related record.
+
+A relation is considered dereferenced if it contains more than one key.
+
+The syntax was chosen for a number of reasons:
+
+- We would like to avoid confusion with JSONRef as it is already implemented in Invenio, and will work differently than JSONRef.
+- We would like a syntax that can be used both in the internal JSON document stored in the database, as well as in the external JSON used in the REST API.
+- By using an object we can easily dereference it. I.e. ``{"id": "..."}`` is dereferenced to ``{"id": "...", "key1": "val1", ...}``. The key ``id`` always exists (unlike e.g. when using JSONRef).
+- Relations should be to a persistent identifier for the related record to make the relation as stable as possible. Also, we wanted to avoid having mulit-key objects as possible persistent identifiers to ensure APIs could be as standardized as much as possible.
+- We did not want to include e.g. type information for the related record, to keep the JSON small as well as keep the type information on the field definition itself, instead of inside the referenced object.
+
+#### Persistent identifiers
+
+We imagine that the persistent identifiers depends on the related record. If the related record uses Invenio-PIDStore, it would be the persistent identifer stored in PIDStore. If the related record just uses the record's table UUID as persistent identifier, this value could be used.
```

Comment:
Not sure if I follow or if this is too much information: the persistent identifier is always an Elasticsearch record key, right? Whether that key represents the record's table UUID or not doesn't matter, no?

Perhaps a better question is: how does the system know where to go to dereference apart from code? Answer: it doesn't. Only the code knows that.